clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -115,7 +115,10 @@ public long append(long offset, long timestamp, byte[] key, byte[] value) {
      * to accept this single record.
      */
     public boolean hasRoomFor(byte[] key, byte[] value) {
-        return this.writable && this.compressor.numRecordsWritten() == 0 ?
+        if (!this.writable)
+            return false;
+
+        return this.compressor.numRecordsWritten() == 0 ?
```

Comment:
I know this is existing but I don't completely understand why we need seperate boolean test cases here:
```scala
return this.compressor.numRecordsWritten() == 0 ?
   this.initialCapacity >= Records.LOG_OVERHEAD + Record.recordSize(key, value) :
   this.writeLimit >= this.compressor.estimatedBytesWritten() + Records.LOG_OVERHEAD + Record.recordSize(key, value);
```

Why is `initialCapacity` used on the first record and `writeLimit` for the rest? I think when `this.compressor.numRecordsWritten() == 0` then `this.compressor.estimatedBytesWritten()` will be 0 too. So perhaps we could simplify this to:

```scala
return this.writeLimit >= this.compressor.estimatedBytesWritten() + Records.LOG_OVERHEAD + Record.recordSize(key, value);
```

Do you have more background @ijuma?